### ISSUE TITLE
uberon and MONDO parser

### ISSUE THREAD

**Issue by pnrobinson** [monarch-initiative/phenol#23]
@cmungall @yy20716 I was wondering if it would be a good idea to develop a MONDO OBO Parser with OLPG using a new "generic" OBO Parse policy. Then, I could introduce MONDO into the HPO Workbench (and a little down the line we could also introduce MONDO into the HPO Webpage).

Would a decent strategy be to declare classes called "GenericTermId" etc etc so that we would not break anything that currently works, and then one say the MONDO is working with the "Generic.." classes, to move HPO, MP, and GO over to the generic classes, and then to remove the other classes and rename "GenericTermId etc to TermId etc.

**Comment by yy20716** [comment]
- If you see my previous commit, you would notice that the bridge codes based on `Obographs` already use generic classes such as `commonTermId/Relation` to internally map nodes/edges in `Obographs` to corresponding ones in OLPG (i.e. the codes were mainly tested over NCIT ones but not specifically designed for NCIT only). I believe it would be able to parse `mondo.owl` as well. The `commonTermId/Relation` are just copies of classes from the ones for GO for now. 
- I will just re-name of the classes as `GenericTermId/Relation` and commit again.

**Comment by pnrobinson** [comment]
@yy20716 thank you, I did see the generic class names but apparently had a brain disconnect - sorry! 
One question about the design of OLPG is whether we want to include classes such as MondoUtility that would have a bunch of commonly used functions (e.g.,
```
TermId  getMondo(String someOtherDatabaseId) {
   TermId mondoId = find Mondo class that corresponds to some OMIM id, for instance;
  return mondoId;
}
```
and possibly functions that would implement clinical categories -- get all neuro diseases for instance.

**Comment by cmungall** [comment]
I like this idea

I would do this as facades over some more general method

so for example a generic lookupByEquivalentClassId, LookUpByXref; also 
getAllInSubset, getAllSubClassOf

then in the mondo convenience layer, getMondoByOmim; getRareSubset; 
getNeuroSubClasses

**Comment by pnrobinson** [comment]
One important missing feature right now is that ontolib does not get you the PMID database_xrefs for the HPO definitions, which would be great to show on the Website (which is probably not realistic for this iteration anyway but will be nice to have in the future). We should definitely add this with the new classes.

**Comment by cmungall** [comment]
@yy20716 here is an example of how definition xrefs are stored in the underlying obographs java:

https://github.com/geneontology/obographs/blob/2140a947a6d11bc68729ae8954b7b3723872cad5/src/test/java/org/geneontology/obographs/model/MetaTest.java#L18-L51

Although it's easier to see in the JSON:

https://github.com/geneontology/obographs/blob/6b9eb853e94425e0a1ceb676d4517b160eff502e/examples/nucleus.json#L8-L13

**Comment by yy20716** [comment]
I am sorry, but I am not sure whether I understood your suggestions. If you don't mind, could you please clarify my questions?

1. Can you give some more details or descriptions on methods you suggested? My guess is that  
    - `lookupByEquivalentClassId`: given a term/class id,  get all equivalent classes. 
    - `getAllSubClassOf`: given a term/class id,  get all subclasses (i.e. all the objects of rdfs:SubclassOf+) 
    - `lookUpByXref`: get all classes where their cross-reference classes match the given class.
    - `getMondoByOmim`: get all mondo classes having the corresponding OMIM ones.

    But I am not sure about the following ones: `getAllInSubset`, `getRareSubset`, `getNeuroSubClasses`. Can you give overall descriptions/specifications of these methods? If these are discussed in other threads, could you please let me know?

2. I also get to know that some general algorithms are already implemented, e.g., under ontology/algo package, it seems that Peter already implemented [OntologyAlgorithm](https://github.com/monarch-initiative/OLPG/blob/master/ontolib-core/src/main/java/com/github/phenomics/ontolib/ontology/algo/OntologyAlgorithm.java), which is an implementation of several commonly needed algorithms for traversing and searching, etc. 
    - I wonder whether I could add the general methods suggested in this `OntologyAlgorithm` (e.g., `lookUpByXref`). 
    - In the same package, do you think I could also add specialized classes for each ontology such as `MondoOntologyAlgorithm` for Mondo? This class would have utility methods such as `getMondoByOmim` implemented using the methods in `OntologyAlgorithm`. 

3. so do "PMID database_xrefs for the HPO definitions" in Peter's comment mean the definition xrefs under HPO's entity/term? So does this mean that we just need a method that retrieves all xrefs under definitions?

**Comment by pnrobinson** [comment]
@yy20716 Maybe it would be good if we could skype about this. I think there will be different requirements for the different ontologies also. But for the HPO we will need the methods in the OntologyAlgorothm class. 
* lookupByEquivalentClassId -- the use case is that many terms have an "alt_id" created by merging two old terms. Currently, OLPG crashes if an alt_id term is used for a query, and so we need to mitigate this (perhaps by translating on the fly)
* lookUpByXref -- yes. I think that OLPG does this already
one new function is "get all the annotations for a definition or synonym or comment" Usually, this will be the biocurator or a PubMed id. In HPO, these are attached as annotations to the "definition" element (always), and sometimes to the rdfs:comment element, and sometimes to synonym elements.

I think this issue is getting overloaded. I think we have decided to use OWLAPI and TInkerPop to move forward? If nobody objects, I would close this issue and make smaller subissues for any open points.

**Comment by yy20716** [comment]
* I guess it would be better if we could discuss with Chris. In my preliminary opinion, it would be nice if we could have a specialized class for HPO such as `HPOOntologyAlgorithm`, i.e. existing algorithms for now can be moved to that class and then we could move some of them back to `OntologyAlgorithm` if they can also be used for other ontologies as well. 
  - Alternatively, we could add new packages such as `generic`, `hpo`, `go`, etc and put algorithm classes like [the organization used in JGraphT](https://github.com/jgrapht/jgrapht/tree/master/jgrapht-core/src/main/java/org/jgrapht/alg).

* I put my comment about TinkerPop in [this post](https://github.com/monarch-initiative/OLPG/issues/6).

* Yes, it would be better if we could discuss specific issues rather than mixing them up. Thank you.

**Comment by pnrobinson** [comment]
Thanks to Chris and HyeongSik for the discussions and insight. I am copying the following here from an email by HyeongSik about where to go with the Mondo parser.
*  refer to testcases in "OwlImmutableOntologyLoaderTest.java" for parsing mondo ontology file. This testcase used a dummy ncit dataset, but it can be used for mondo ontology. Please note that
 - This loader is built based on Obographs, so the mappings from owl to its internal graph structure also follows the mappins in Obographs.
 - This loader only loads a basic graph structure with some basic mappings such as synonyms. In the case where you need additional mappings such as equivalent classes, domain/range axioms in mondo, please let us know. Thank you.

I would suggest that we implement the parser as above in HPOworkbench, and if we are satisfied, then we write detailed documentation about how to use this kind of code for Mondo parsing in the RTD documentation.

**Comment by pnrobinson** [comment]
@yy20716 

I have added a protoype Mondo parser to HpoWorkbench in the branch phenol_1_8
https://github.com/TheJacksonLaboratory/HPOworkbench/tree/phenol_1_8

(Note that we should not merge this just yet because it is not possible to download the new annotation format yet -- we need to find a home for phenote.hpoa before we can update HPOworkbench)

In any case, the basic code is just this
```
public ImmutableOntology<GenericTerm,GenericRelationship> parse() throws PhenolException{
        final OwlImmutableOntologyLoader<GenericTerm, GenericRelationship> loader =
                new OwlImmutableOntologyLoader<GenericTerm, GenericRelationship>(
                        new File(pathToMondoFile));

        final GenericOwlFactory cof = new GenericOwlFactory();
        try {
            final ImmutableOntology<GenericTerm, GenericRelationship> ontology = loader.load(cof);
            return ontology;
        } catch (OWLOntologyCreationException | IOException e) {
            throw new PhenolException(String.format("Could not create Mondo ontology: %s",e.getMessage() ));
        }
    }
```
but this leads to the following error (run the download command first to get the mondo.obo file)

```
$ java -jar hpowbcli/target/HPOWorkbench.jar mondo
TRACE [main] (HpoWorkbench.java:19) - running command mondo
TRACE [main] (MondoCommand.java:36) - running Mondo command
ERROR - 2018-03-20 17:47:46 - /home/robinp/IdeaProjects/HPOworkbench/src/main/resources/curie_map.yaml (No such file or directory)
java.io.FileNotFoundException: /home/robinp/IdeaProjects/HPOworkbench/src/main/resources/curie_map.yaml (No such file or directory)
```

I am not sure where the curie_map.yaml is coming from?

**Comment by yy20716** [comment]
I am sorry about this error. Phenol currently uses that curie_map.yaml to maintain mappings between CURIEs (compact IRIs) and IRIs. I borrowed the file from [Dipper](https://github.com/monarch-initiative/dipper/blob/master/dipper/curie_map.yaml) for that purpose but the same file is also available in [phenol-io's main/resources](https://github.com/monarch-initiative/phenol/tree/master/phenol-io/src/main/resources). Anyway, it seems that the phenol's parser codes was looking for main/resources directory which was the one in HPOworkbench's directory. I will see how I can solve this issue. Thank you.

**Comment by pnrobinson** [comment]
This seems to be working now.
